proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,318 @@
+
+Haskell users have different tolerances for risks related to language change. Some derive great value from using the newest features, while others are conservative. GHC should serve the whole spectrum. Similarly, users typically do not revisit the documentation for a language feature unless the compiler's behavior surprises them, so users should be able to specify their extension use policies and then receive feedback from the compiler when their code violates said policies, whether this violation is due to changes in the code or changes in the compiler.
+
+There will be an additional set of warnings:
+ * `-WXDeprecated`: Issue a warning when a deprecated extension is enabled
+ * `-WXStable`: Issue a warning when a stable extension is enabled
+ * `-WXExperimental`: Issue a warning when an experimental extension is enabled 
+ * `-WXLegacy`: Issue a warning when a legacy extension is enabled.
+
+Each category will also support the usual`-Wno-` syntax, so `-Wno-XDeprecated` will turn off warnings for deprecated extensions. Additionally, each extension should have a configurable warning that can be individually enabled or disabled, allowing users and teams to inform GHC about their local policies and needs. For instance, `-WXDeprecated -Wno-XRank2Types` would warn about all deprecated extensions except `Rank2Types`.
+
+Furthermore, `-WXDeprecated` and `-WXExperimental` would be added to `-Wcompat` for the next release to allow the community time to adjust, and then they will be on by default. This is because these two categories describe extensions that are likelier to lead to incompatibilities with future releases of GHC. `-WXLegacy` should be added to `-Wall`, so that only those who request extra feedback will get it.
+
+Note that this warning syntax rules out the names `Deprecated`, `Stable`, `Experimental`, and `Legacy` for future language extensions.
+
+#### 2.2.1. Warnings are for Enabled Extensions
+
+Extension warnings will be issued when an extension is enabled, not when the language features that it supports are used in a program. The following program should issue a warning, even though no multi-way conditionals occur in it:
+```
+{-# OPTIONS_GHC -WXMultiWayIf #-}
+{-# LANGUAGE MultiWayIf #-}
+module Main where
+
+main :: IO ()
+main = pure ()
+```
+
+There are two main reasons to do this, rather than wait for the associated code path in the compiler to be invoked:
+
+ 1. Users don't need to think about when one extension implies features from another. The mental model for warnings is simple and straightforward, and it's easier to see the connection between a change in the code and a new warning in CI.
+ 
+ 2. The implementation is likely to be simpler, because the extension lifecycle warnings can presumably be issued without needing to scatter them throughout the parser and type checker.
+
+Warnings should be issued for extensions no matter how they are enabled. For instance, the following program should also issue a warning:
```

Comment:
What location is used for the warning? It is clear where to put it when there is a language pragma, but not so here. And presumably we don't want to emit a warning for every module that GHC compiles in this case? Or do we? Perhaps it should be like `-Wunused-packages`, and not have a location.

Reply:
Should the specific source locations of warnings belong here? I saw this as something like the specific wordings - up to the one doing the implementation.

Off the cuff, I'd think that the best possibility would be to put as many as possible on `LANGUAGE` or `OPTIONS_GHC` pragmas, but when that can't happen, either do it locationless or on the `module` header at the top if there is one? But this seems like the kind of thing that the implementers should decide, and adjust as needed.

Reply:
I agree we don't need to pin down the implementation too heavily regarding source locations. But it's a good question as to whether warnings are issued once per module or not. This section as written implies they are, so e.g. adding an experimental extension to `default-extensions` would result in a warning on every module in the program. Is that what we want, when arguably the location we'd like to point the user to is in the `.cabal` file?

It is pretty clear that we should report a per-module warning on any `LANGUAGE` or `OPTIONS_GHC` pragma that enables a warned-about extension. But for extensions that are not explicitly enabled in the file, rather than each module reporting duplicate warnings for extensions that were enabled globally, perhaps we should have the driver report them once?

I'm in two minds about this. One consequence would be that in the following example, we wouldn't get a warning, because `-WXBangPatterns` is local to the file whereas `BangPatterns` was enabled globally:
```hs
{-# OPTIONS_GHC -WXBangPatterns #-}
module Main where
```

Reply:
I guess it seems to me that the location is part of the UX of this feature, and we _have_ been discussing UX a bit in this proposal.

Reply:
Thanks for the detailed response @david-christiansen, it's fairly convincing that one-module-at-a-time is a reasonable way forward.

> The second difficulty is that if the `.cabal` file turns on an experimental extension, but each module has a pragma that turns it off, then the extension is in fact not enabled anywhere. As a user, I'd expect the warning to not occur until I create a file without the pragma that turns it off.

What about `ghci`/`cabal repl`? Presumably it should emit warnings about experimental extensions, even in the absence of any modules?

Reply:
> What about ghci/cabal repl? Presumably it should emit warnings about experimental extensions, even in the absence of any modules?

In fact, I think there are in principle a whole class of awkward cases (although this is getting quite far off-topic). Suppose that, for example, the GHC team wanted to remove a flag from GHC. Arguably, it would be nice to emit a deprecation warning _about the flag_ for a release or two before removing it. But this has exactly the same problems that we've been discussing (since setting extensions is kind of a subset of setting a flag), but it's harder to avoid the "no location" case.

Thinking of it this way does make me think slightly differently about some of David's analysis.

> The second difficulty is that if the .cabal file turns on an experimental extension, but each module has a pragma that turns it off, then the extension is in fact not enabled anywhere. As a user, I'd expect the warning to not occur until I create a file without the pragma that turns it off.

If we consider the case of deprecated extensions, _this is still problematic_! Because if the extension gets removed, your build will still break because GHC will reject the flag despite it never being removed. So that makes me think that you _should_ get a warning here.

This makes me suggest a slightly extended version of the simple approach that David suggested, but this time extending to _anything that sets a flag_, extension or otherwise.
- A warning is emitted _when the flag is enabled_, regardless of whether it is later disabled or whether it will not take effect for some other reason. The "warnable event" is turning on the flag, that's it (easy to understand!)
- The warning is emitted where the flag is set, if possible:
    - For pragmas it's easy
    - For command line flags (or e.g. `:set` in GHCi), we emit a no-location warning

This doesn't have the best UX, but it's mechanically relatively simple (I think) and easy for users to understand why it's doing what it's doing in the end.

Reply:
> A warning is emitted when the flag is enabled, regardless of whether it is later disabled 

Won't this be awkward in the case of `-XGHC20xy -XNoDeprecatedExtensionPartOfGHC20xy`?

Reply:
I think that Joachim's concern here should be taken seriously. Enabling a set and then disabling a few elements of it is a common practice. We shouldn't train people to turn off the warning by giving them false positives.

How about this as a loose specification: a warning is emitted if GHC ever enters a state where it will accept code when the warnable extension is enabled? These states are, so far as I can see, when a prompt becomes available at GHCi or when a module header has been parsed.

This seems to cover language pragmas, options pragmas, command-line flags, GHCi, and all the other situations I can think of. At the same time, it's short and uncomplicated, and doesn't deliver false positives.

Reply:
> Won't this be awkward in the case of -XGHC20xy -XNoDeprecatedExtensionPartOfGHC20xy?

If a part of GHC20xy is deprecated, then it is going to be removed in the future, perhaps breaking GHC20xy entirely? In this situation it seems like maybe I should be worried.

Personally I'd be willing to bite this bullet but I can see why you might not want to.

Reply:
Maybe I should have said `-XNoLegacyExtensionButPartOfGHC20xy`.

Or a setting `-WXBenignExtensionButUnwantedInThisOrg` followed by `{-# LANGUAGE GHC2023 #-}{-# LANGUAGE NoBenignExtensionButUnwantedInThisOrg #-}`

Reply:
I guess the thing I'm gesturing at is that for _removal_, the thing that is at risk of breakage is not just your _code_ but also your _flag specification_. If the extension `Foo` is removed, then my build which says `-XFoo -XNoFoo` _will break_.

Maybe we're fine with that! Perhaps the scope of this is really just the source code. After all, we don't currently tend to offer deprecation periods or warnings for flags that get removed. But maybe we should!

Reply:
I think that this suggests something that's out of scope of this proposal, but that would be a very sensible policy: if an extension is removed, disabling it should be a no-op rather than an error. In other words, lets say for the sake of argument that we remove the extension `AlternativeLayoutRuleTransitional` in the future. Code that specifies `{#- LANGUAGE NoAlternativeLayoutRuleTransitional #-}` should continue to work indefinitely, as should code that is built with `-XAlternativeLayoutRuleTransitional` but then has that language pragma in the header to turn it off. The "invalid extension" error for extensions that have been removed should only occur when GHC is in a state where it will accept Haskell code, and when the extension is in fact enabled.

I generally think that backwards-incompatible changes shouldn't be happening, but there's always a cost to keeping things around, and it seems plausible that a language extension will be removed at some point in the future, after providing a good migration period and fixing up Hackage.

This means of handling formerly-valid extensions should be possible to build in one place, test thoroughly, and document well, and it future-proofs the process as well as possible. It also won't break the builds in the way @michaelpj describes.

Right?

Reply:
This all sounds good to me, and I think could in fact be specified in this proposal as a new lifecycle category: an extension can be `PiningForTheFjords`, at which point GHC will allow it to be disabled as you describe. And there's an expected transition `Deprecated` -> `PiningForTheFjords`.

Although perhaps it would be enough to make an unrecognised extension a warning rather than an error. For some reason `-XScopedWombatPunning` is currently a fatal error, whereas an unrecognised warning flag such as `-Wscoped-wombat-puns` merely issues a warning.

Reply:
> Although perhaps it would be enough to make an unrecognised extension a warning rather than an error.

FWIW I posted a GHC ticket proposing this behaviour: https://gitlab.haskell.org/ghc/ghc/-/issues/23858

Reply:
I've gone back and forth about a "removed" category for this proposal, as you suggest here. It has some nice benefits:
 1. It provides a natural way to have the User's Guide describe why some old program no longer works (the feature was removed in version Y, for this reason, here's how to migrate)
 2. It provides a good way for the compiler to direct users to replacement features or to said documentation page
 3. It allows typos in language feature names to be more than warnings :-)

On the other hand, it seems that the language of pragmas, feature flags and other configurations that GHC accepts has a more general problem with breaking changes, and that language extension flags are somehow a special case of that problem. I've never been a power user of GHC flags, so I'm a poor person to propose a specific plan, but some kind of migration/lifecycle system for compiler flags and options seems useful and important.

I know I'm entering the typical Haskell-culture failure mode here of rejecting a solution that works in favor of a nonexistent theoretical one that could be better in principle, so I'm definitely willing to revise this conclusion.